### ISSUE TITLE
On-screen keypad improvements

### ISSUE THREAD

**Issue by sspanak** [sspanak/tt9#240]
#159 added experimental support for touchscreen-only phones, but it was clear there would be a second phase of small improvements or eliminating things that cause annoyance

Here is the compilation:
- ~~Add suggestion control keys in a new row, right below the suggestion bar: Filter/Clear filter, Select Previous/Next (after #201 , when there will be separate functions for filtering)~~
- ~~Add proper support for the "upside-down keyboard" setting.~~ - fixed in #432 
- ~~Add "Key Hints" Setting. When enabled, assigned hardware keys would be displayed above the software key icons (or labels)~~ - won't do this for the function keys, because there will be no room to put the secondary function (the function when holding).
- ~~The punctuation keys must change to: `#` and `*` when `ModeDialer` is on and `,` and `.`, when `Mode123` is on.~~ - done in #358 
- ~~LEFT and RIGHT virtual keys for scrolling suggestions must move the cursor to the respective direction, when there are no suggestions.~~ - in #546
- ~~Replace the Next Mode button icon with this: ⌨~~ - in #546
- ~~Disable the inactive keys in each mode~~ - in #546
  - ~~123: Disable Change Language, Filtering keys and Add Word~~
  - ~~ABC: Disable Filtering keys~~
  - ~~Voice: Disable all maybe? Figure out what makes sense.~~
- ~~Vibration on key press (+ a setting)~~ - in #546
- ~~Ensure all keys with icons have text equivalents. Settings and Mode do not appear properly in Android <= 5.~~ - in #546
- ~~Make sure the text equivalents have the correct size. For example, "Del" appears oversized on Android 5.~~ - in #546
- ~~Make it possible to resize vertically.~~ - in #551 
  - ~~Add a "Height" settings on the Appearance screen.~~
  - ~~If possible use the status bar as a handle. Dragging it all the way down, will switch to the smaller layout.~~
- ~~Add support for fullscreen/landscape mode. Make sure the we don't crash when the free space is narrow.~~
- ~~Remove the "BETA" label.~~

Numeric field bugs:
- ~~The automatic switching to 123 mode does not trigger key label update~~ - fixed in #265 
- ~~In float fields, typing "," and "." does not work. Punctuation keys do nothing when held.~~ Fixed in #255

**Comment by Teloshav** [comment]
The main page says the following:
> For touchscreen-only devices, an on-screen keypad can be enabled in the Settings.

After reading some of the issues you've opened about implementing this feature, I haven't been able to get this software keyboard to display at all...

I'm almost certain I'm doing something wrong, but I'm not sure where exactly I should ask about it.

I am using a Google Pixel 6 with GrapheneOS.

- Teloshav

**Comment by Teloshav** [comment]
Sounds great! Can't wait to try it out. I'll be looking here every now and again until version 21 comes out.

Is there a way I could get a notification for when this new version comes out? (I am extremely new to Github.)

- Teloshav

**Comment by sspanak** [comment]
>  I'll be looking here every now and again until version 21 comes out.

@Teloshav, the new version is out. Enjoy!

> Is there a way I could get a notification for when this new version comes out? (I am extremely new to Github.)

You could give the repo a star or start watching it using the buttons on top and you will start receiving emails about all activity here.

**Comment by gabriel-tandil** [comment]
I celebrate the addition of the virtual keyboard. It really gives a new life to the project. ❤️

It would be nice if the numbers were somewhat more noticeable than the letters (as in the physical keyboards, I think it would be more readable) and also could hide the letters with accents to help in this regard.

Also the total size of the virtual keyboard could be smaller to not occupy so much screen.

**Comment by sspanak** [comment]
> Also the total size of the virtual keyboard could be smaller to not occupy so much screen.

It definitely does. I am thinking of a way of rearranging things on screen without taking away any functionality.

> ... and also could hide the letters with accents (in the languages that have them) 

Well, it is hard to determine which letters are supposed to be hidden. Currently, the virtual keys are configured to display the first 5 characters. This is enough for the languages that have many different letters for a single key and to provide a hint you could type letters with accents on languages with less letters, but having accented ones. I haven't come up with a way of displaying only "the necessary" ones without overcomplicating the code. If you have an idea how to do it in a simple way, please, go ahead and suggest it.

**Comment by gabriel-tandil** [comment]
> Well, it is hard to determine which letters are supposed to be hidden

I think that in all cases it could be determined by the character code, since accented letters have higher codes than unaccented ones, but we would have to see if in any language with non-Latin characters this is also true. Otherwise, a (less satisfactory) possibility could be to show fixed the standard English keyboard layout, which almost everybody is used to, even if they do not type in English.

I would like to collaborate with the development and some spanish translations that are missing, I should build myself an environment to develop in android that currently I don't have available.

**Comment by sspanak** [comment]
> I think that in all cases it could be determined by the character code, since accented letters have higher codes than unaccented ones, but we would have to see if in any language with non-Latin characters this is also true.

Actually, that's a very good idea. I can certainly filter out the "latin+accents" range and leave out the rest. And as of now, all other languages do not have accented letters, so we can leave them be.

> I would like to collaborate with the development and some spanish translations that are missing, I should build myself an environment to develop in android that currently I don't have available.

Oh, please do. Translations are the most difficult part for me, as you can imagine. If it's only the translations, you could probably get away with only editing the respective `values.xml` file in any text editor. I've set up a CI to do some simple code validation of the code, the pull request will shine in red. It is far from perfect, but it is something.

And if you want to set up Android Studio, it should be fairly simple. Just use the latest version and import the project in it. A couple more configuration steps are described [in the Contribution Guide](https://github.com/sspanak/tt9/blob/master/CONTRIBUTING.md), then you should be good to go.

**Comment by BUCK113** [comment]
New to here but loving the keyboard so far but few suggestions, Ive got big thumbs so being able to adjust keyboard size would be great,
having a long press function to select individual letters or numbers. Having a caps lock key for cap letters. Combine exclamation and question to have a larger space bar, maybe having a oment bar to free up keyboard space to be able to utilize keyboard space. Just my humble opinion

**Comment by maverick74** [comment]
Question: is this bug a requirement to get the on-screen keypad out of the beta label?

I ask this because beta implies a testing state that is subject to frequent bugs and a still not ready for people in general!

For the past months I've been using it, that does not seem to be the case anymore! I think it's pretty stable! (yeah, occasionally there are bugs, but the same is true for software out of beta)!

So I don't know if there is any profit in keeping it labeled as beta unless there is other reasons than bugs...

**Comment by maverick74** [comment]
Well, the right and left arrows for moving the cursor is more of a feature than a bug! (yeah, I would like that too!!!)

Now about the broken icons and the oversized text... that do seem like an issue worthy of the "bug" label.
You seem to already have a nice solution for the icon, however... and I suppose the oversized text you mentioned is where there is no alternative icon and you have to use text instead... (if that's the case, it's a weird behavior! Why doesn't it follow the size set for the other keys?!)

in any case, it does seem you already have a plan (I guess you're just lacking time for it!!! Lol XP)

**Comment by sspanak** [comment]
>  I suppose the oversized text you mentioned is where there is no alternative icon and you have to use text instead... (if that's the case, it's a weird behavior! Why doesn't it follow the size set for the other keys?!)

It's because the icons are actually emojis. And you are correct there is a fallback text when the specific emoji is not available on a given phone. The problem is, I have set a slightly larger font size for the emojis for better readability, but I didn't check if the fallback text looks OK. That's all! It shouldn't be difficult to fix, I just need to find the time for that.

> in any case, it does seem you already have a plan (I guess you're just lacking time for it!!! Lol XP)

Exactly!
